[Fix #95] Prevent crash of cron component

sebastianpoeplau opened the pull request:
Additionally, add a test for the component.

Commit:
[Fix #95] Prevent crash of cron component

Additionally, add a test for the component.

prayerslayer commented:
Finally someone caring about `cron`, the ugly duck of friboo :grin:

prayerslayer commented:
:+1:

harti2006 commented:
:+1:

dryewo commented on test/org/zalando/stups/friboo/system/cron_test.clj:
```diff
@@ -0,0 +1,16 @@
+  ;; Here we make sure that the component is started and stopped properly.
+  (let [state (promise)
+        cron-component (map->TestCron {:state state})]
+    (.start cron-component)
```
I suspect that this does not test anything, as `cron-component` is immutable and not started in the beginning.
The right way to do it should be something like:

``` clj
(-> (map->TestCron {:state state})
    (component/start)
    (component/stop)
    (component/stop))
```

so that each following call uses the result of the previous one.

sebastianpoeplau replied:
Good point, will change it. Actually, the test was enough to reproduce the crash, but you're right in saying that it doesn't actually stop the running component.

Commit:
Fix unit test

The Cron component is immutable, so the right approach is to thread it
through the calls to `start` and `stop`.

Commit:
Fix race condition in unit test

Multithreading is tricky... :/

dryewo commented:
👍

sebastianpoeplau commented:
👍

sebastianpoeplau commented:
Too bad, I thought I might be able to give the second approval ;) Is it because I'm not in the Zalando group?

sebastianpoeplau commented:
Okay, I've been added to the org. Now my approval should count :)

sebastianpoeplau commented:
👍

sebastianpoeplau commented:
I don't seem to be authorized to merge...
